arcgis-ios-sdk-samples/Display information/Create symbol styles from web styles/CreateSymbolStylesFromWebStylesViewController.swift

```diff
@@ -0,0 +1,308 @@
+            // For each category value of a symbol, we need to create a
+            // unique value for it, so the field name matches to all categories.
+            category.symbolCategoryValues.map { symbolValue in
+                AGSUniqueValue(description: "", label: category.symbolName, symbol: symbol, values: [symbolValue])
+            }
+        }
+        return renderer
+    }
+    
+    // MARK: UIViewController
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Add the source code button item to the right of navigation bar.
+        (navigationItem.rightBarButtonItem as? SourceCodeBarButtonItem)?.filenames = ["CreateSymbolStylesFromWebStylesViewController"]
+        
+        // Get the symbols from the symbol style hosted on an ArcGIS portal.
+        getSymbols(symbolStyle: symbolStyle, categories: SymbolCategory.allCases) { [weak self] in
+            guard let self = self else { return }
+            self.featureLayer.renderer = self.makeUniqueValueRenderer(fieldNames: ["cat2"])
+            self.legendBarButtonItem.isEnabled = true
+        }
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        mapScaleObservation = mapView.observe(\.mapScale, options: .initial) { [weak self] (_, _) in
+            DispatchQueue.main.async {
+                guard let self = self else { return }
+                self.featureLayer.scaleSymbols = self.mapView.mapScale >= 8e4
+            }
+        }
+    }
+    
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+        mapScaleObservation = nil
+    }
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == "LegendTableSegue",
+           let controller = segue.destination as? UITableViewController {
+            controller.presentationController?.delegate = self
+            // The data source for the legend table display.
+            symbolsDataSource = SymbolsDataSource(tableView: controller.tableView, symbols: symbols)
+            controller.tableView.dataSource = symbolsDataSource
+        }
+    }
+    
+    func adaptivePresentationStyle(for controller: UIPresentationController, traitCollection: UITraitCollection) -> UIModalPresentationStyle {
+        // Ensure that the settings are shown in a popover on small displays.
+        return .none
+    }
+}
+
+// MARK: - SymbolsDataSource
+
+private class SymbolsDataSource: NSObject {
+    /// A weak reference to the table view which the data source itself connects.
+    private weak var tableView: UITableView?
+    
+    /// The symbols created from a symbol style.
+    var symbols: [(category: SymbolCategory, symbol: AGSSymbol)] {
+        didSet {
+            // Reset caches to ensure legends are updated correctly.
+            resetCaches()
+            tableView?.reloadData()
+        }
+    }
```

Comment:
If you remove `CreateSymbolStylesFromWebStylesViewController.symbols`'s property observer, then this property can be made immutable:

```swift
let symbols: [(category: SymbolCategory, symbol: AGSSymbol)]
```

Makes `resetCaches()` unnecessary. This was made possible by changing to use a custom data source rather than a custom table view controller.